core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -99,23 +104,71 @@ class CloudSqlInstance {
     }
 
     synchronized (instanceDataGuard) {
-      this.currentInstanceData = executor.submit(this::performRefresh);
+      this.currentInstanceData = this.startRefreshAttempt();
       this.nextInstanceData = currentInstanceData;
     }
   }
 
   /**
-   * Returns the current data related to the instance from {@link #performRefresh()}. May block if
-   * no valid data is currently available.
+   * Returns the current data related to the instance from {@link #startRefreshAttempt()}. May block
+   * if no valid data is currently available. This method is called by an application thread when it
+   * is trying to create a new connection to the database. (It is not called by a
+   * ListeningScheduledExecutorService task.) So it is OK to block waiting for a future to complete.
+   *
+   * <p>When no refresh attempt is in progress, this returns immediately. Otherwise, it waits up to
+   * 30 seconds. If a refresh attempt succeeds, returns immediately at the end of that successful
+   * attempt. If no attempts succeed within the 30 second timeout, throws a RuntimeException with
+   * the exception from the last failed refresh attempt as the cause.
+   *
+   * <p>The behavior of this method has changed from earlier releases. In version 1.14.1 and
+   * earlier, When no refresh attempt is in progress, returns immediately. Otherwise, blocks
+   * application thread until the current refresh attempt finishes. If the refresh attempt succeeds,
+   * this returns the InstanceData. If not, this throws a RuntimeException, while a new refresh
+   * attempt is submitted to the executor in the background.
    */
-  private InstanceData getInstanceData() {
+  private InstanceData getInstanceData(long timeoutMs) {
     ListenableFuture<InstanceData> instanceDataFuture;
     synchronized (instanceDataGuard) {
       instanceDataFuture = currentInstanceData;
+      // If the currentInstanceData has expired, then force refresh (which will balk if a refresh
+      // is already running) and make this and future requests to getInstanceData wait on the
+      // refresh operation to complete.
+      if (instanceDataFuture.isDone()) {
+        Instant expiration;
+        try {
+          expiration = instanceDataFuture.get().getExpiration();
+          if (expiration == null || expiration.isBefore(Instant.now())) {
```

Comment:
If the current instance data has already expired, won't we ended up initiating a force refresh when the socket creation fails? In other words, do we really need to do this work here?